api/src/main/java/org/apache/iceberg/transforms/ProjectionUtil.java

```diff
@@ -52,6 +52,84 @@ private ProjectionUtil() {}
     }
   }
 
+  static <T> UnboundPredicate<T> truncateIntegerStrict(
+      String name, BoundPredicate<Integer> pred, Transform<Integer, T> transform) {
+    int boundary = pred.literal().value();
+    switch (pred.op()) {
+      case LT:
+        // Checking if the literal is at the lower partition boundary
+        if (transform.apply(boundary - 1).equals(transform.apply(boundary))) {
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary - 1));
+        } else {
+          return predicate(Expression.Operation.LT_EQ, name, transform.apply(boundary - 1));
+        }
+      case LT_EQ:
+        // Checking if the literal is at the upper partition boundary
+        if (transform.apply(boundary + 1).equals(transform.apply(boundary))) {
+          return predicate(Expression.Operation.LT, name, transform.apply(boundary));
+        } else {
+          return predicate(Expression.Operation.LT_EQ, name, transform.apply(boundary));
+        }
+      case GT:
+        // Checking if the literal is at the upper partition boundary
+        if (transform.apply(boundary + 1).equals(transform.apply(boundary))) {
+          return predicate(Expression.Operation.GT, name, transform.apply(boundary + 1));
+        } else {
+          return predicate(Expression.Operation.GT_EQ, name, transform.apply(boundary + 1));
+        }
+      case GT_EQ:
+        // Checking if the literal is at the lower partition boundary
+        if (transform.apply(boundary - 1).equals(transform.apply(boundary))) {
+          return predicate(Expression.Operation.GT, name, transform.apply(boundary));
+        } else {
+          return predicate(Expression.Operation.GT_EQ, name, transform.apply(boundary));
+        }
+      case NOT_EQ:
+        return predicate(Expression.Operation.NOT_EQ, name, transform.apply(boundary));
+      default:
```

Comment:
It would be good to add an explicit case for `EQ`:

```
case EQ:
  return null; // there is predicate that guarantees equality because adjacent ints transform to the same value
```

Reply:
Ack